### PR TITLE
Add iter ops

### LINKED ISSUE #57: Set operations for multiple maps at a time
Hey,

I discovered the pure Rust version of the Roaring library of Daniel Lemire et al., performances are pretty good at least way better than [my `sdset` library](https://github.com/Kerollmops/sdset).

I was wondering if you had any rough idea of how to implement a set operation and more specifically an union on multiple set at a time.

I dived a little bit into the C implementation and found out that the main performance gain seems to be related to the fact that the library doesn't change the internal representation (and keep the bitmap one) until all sets operations have been applied, then the internal type is changed. Am I right?

https://github.com/RoaringBitmap/CRoaring/blob/59d70d010da5f606f1339fb4c4f200be11f590c6/src/roaring.c#L611-L629

Thank you for your great job!

### PR DESCRIPTION
**Pull request by saik0**
Closes #58
Closes #109
Fixes #57

Adds operators to iterators of bitmaps

# TODO
 - [ ] Treemap ops
 - [ ] Tests
 - [ ]  Examples in docs

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -45,3 +45,38 @@ impl fmt::Display for NonSortedIntegers {
...
 impl Error for NonSortedIntegers {}
+
+/// An [`IntoIterator`] blanket implementation that provides extra methods for [`RoaringBitmap`]
+/// and [`RoaringTreemap`]
+///
+/// This trait is parameterized by a sealed trait and cannot be implemented for types outside
+/// of this crate
+pub trait IterExt<T>: IntoIterator<Item = T>
+where
+    T: private::Roaring,
```
Comment (Kerollmops) [inline src/lib.rs:56]: I am not sure why you want to restrict this trait to our library types? What's dangerous in implementing that on, for example, a wrapper of `RoaringBitmap/Treemap`?

File: src/lib.rs [hunk 2]
```diff
@@ -45,3 +45,38 @@ impl fmt::Display for NonSortedIntegers {
 }
 
 impl Error for NonSortedIntegers {}
+
+/// An [`IntoIterator`] blanket implementation that provides extra methods for [`RoaringBitmap`]
+/// and [`RoaringTreemap`]
+///
+/// This trait is parameterized by a sealed trait and cannot be implemented for types outside
+/// of this crate
+pub trait IterExt<T>: IntoIterator<Item = T>
```
Comment (Kerollmops) [inline src/lib.rs:54]: Maybe we could find a better name for this trait, as it is a trait that provides set operations on multiple sets at the same time, maybe `MultiOps` or something like that? What do you think?

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (saik0) [reply src/lib.rs:56]: Adding new non-defaulted methods would be a breaking change downstream. I was thinking this was for public API compatibility.

File: src/lib.rs (hunk 2 above)
Comment (saik0) [reply src/lib.rs:54]: Sure. We could also create a Roaring prelude that includes iter extensions, RoaringBitmap, RoaringTreemap

**Comment by saik0** [comment]
I see value in exposing an API for lazy ops regardless of whether or not we expose the iterator extensions that use them internally. To generalize to any kind of aggregation the user might need to do.

Thinking out loud in pseudo-rust about the _shape_ of the API, not the names.
```rust
struct Lazy {
  bitmap: RoaringBitmap
}

impl BitOrAssign<&Lazy> for Lazy {
 // .. do a lazy or
}

impl BitXorAssign<&Lazy> for Lazy {
 // .. do a lazy xor
}

impl RoaringBitmap {
  fn into_lazy(self) -> Lazy
}

impl Lazy {
  /// equivalent of repairAfterLazy in C, Java (Go?)
  fn into_bitmap(self) -> RoaringBitmap
}

fn main() -> {
  // Roughly equivalent to current IterExt (this does not use Cow)
  let bitmaps: Vec<RoaringBitmap> = todo!();
  bitmaps.into_iter().map(RoaringBitmap::into_lazy).reduce(|acc, other| { acc |= other; acc }).into_bitmap();

  // Now do it in parallel!
  use rayon::prelude::*;
  let more_bitmaps: Vec<RoaringBitmap> = todo!();
  bitmaps.into_par_iter().map(RoaringBitmap::into_lazy).reduce(|acc, other| { acc |= other; acc }).into_bitmap();
}
```

I'd need to think more about what it might look like for borrowed bitmaps. 🤔

File: src/lib.rs (hunk 2 above)
Comment (Kerollmops) [reply src/lib.rs:54]: Hum... not sure we have enough types and traits for now for that to be interesting, but maybe later with more types.

**Comment by Kerollmops** [comment]
> To generalize to any kind of aggregation the user might need to do. The intuition is to have an API that allows users to do lazy operations with a newtype, that gets "repaired" (to use the same term as other lang impls) when it's unwrapped.

I like the idea, but we must absolutely make sure to document it enough. That's not clear what a `Lazy` type does. I like the idea of wrapping the bitmaps instead of methods to do lazy operations on them. I like the `Lazy` wrapping type as it look a lot like [the `Wrapping` struct of the std](https://doc.rust-lang.org/std/num/struct.Wrapping.html).

**Pull request closed without merging**